File: cli/src/port-forward.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { spawn, type ChildProcess } from 'child_process';
import chalk from 'chalk';
import { input, select, confirm } from '@/prompts.js';
import {
  pickManagedContainer,
  containerWorkspace,
  listAllContainers,
  statusLabel,
  type DockerContainer,
} from '@/container-picker.js';

export interface PortForwardConfig {
  portMapping?: string;
  alias?: string;
  service?: string;
  help: boolean;
  interactive: boolean;
}

export interface ParsedPortMapping {
  localPort: number;
  targetHost: string;
  containerPort: number;
}

export interface PortPair {
  localPort: number;
  containerPort: number;
}

interface PlannedTunnel {
  localPort: number;
  containerPort: number;
  targetHost: string;
  alias: string;
  containerName: string;
  isDevcontainer: boolean;
}

export function parsePortForwardFlags(argv: string[]): PortForwardConfig {
  const flags: PortForwardConfig = {
    help: false,
    interactive: true,
  };

  const positional: string[] = [];

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const next = () => {
      if (i + 1 >= argv.length) {
        throw new Error(`Missing value for flag: ${a}`);
      }
      return argv[++i];
    };
    switch (a) {
      case '-h':
      case '--help':
        flags.help = true;
        break;
      case '--alias':
        flags.alias = next();
        break;
      case '--service':
        flags.service = next();
        break;
      case '--no-interactive':
      case '--non-interactive':
        flags.interactive = false;
        break;
      default:
        if (a.startsWith('-')) {
          throw new Error(`Unknown flag: ${a}`);
        }
        positional.push(a);
    }
  }

  if (positional.length > 1) {
    throw new Error(`Duplicate or invalid arguments: ${positional.join(', ')}`);
  } else if (positional.length === 1) {
    flags.portMapping = positional[0];
  }

  return flags;
}

export function parsePortMapping(mapping: string, defaultService?: string): ParsedPortMapping {
  const parts = mapping.split(':').map((s) => s.trim());

  if (parts.length === 1) {
    const port = parseInt(parts[0], 10);
    if (isNaN(port) || port <= 0 || port > 65535) {
      throw new Error(`Invalid port: ${parts[0]}`);
    }
    return {
      localPort: port,
      targetHost: defaultService || 'localhost',
      containerPort: port,
    };
  } else if (parts.length === 2) {
    const local = parseInt(parts[0], 10);
    const container = parseInt(parts[1], 10);
    if (isNaN(local) || local <= 0 || local > 65535) {
      throw new Error(`Invalid local port: ${parts[0]}`);
    }
    if (isNaN(container) || container <= 0 || container > 65535) {
      throw new Error(`Invalid container port: ${parts[1]}`);
    }
    return {
      localPort: local,
      targetHost: defaultService || 'localhost',
      containerPort: container,
    };
  } else if (parts.length === 3) {
    const local = parseInt(parts[0], 10);
    const targetHost = parts[1];
    const container = parseInt(parts[2], 10);

    if (isNaN(local) || local <= 0 || local > 65535) {
      throw new Error(`Invalid local port: ${parts[0]}`);
    }
    if (!targetHost) {
      throw new Error(`Invalid target host: empty string`);
    }
    if (isNaN(container) || container <= 0 || container > 65535) {
      throw new Error(`Invalid container port: ${parts[2]}`);
    }

    if (defaultService && defaultService !== targetHost) {
      throw new Error(`Conflicting target hosts: mapping specifies '${targetHost}' but --service is '${defaultService}'`);
    }

    return {
      localPort: local,
      targetHost,
      containerPort: container,
    };
  } else {
    throw new Error(`Invalid port mapping format: ${mapping}`);
  }
}

function parsePort(value: string, label: string): number {
  const n = parseInt(value, 10);
  if (isNaN(n) || n <= 0 || n > 65535) {
    throw new Error(`Invalid ${label}: ${value}`);
  }
  return n;
}

export function parsePortPair(item: string): PortPair {
  const parts = item.split(':').map((s) => s.trim());
  if (parts.length === 1) {
    const port = parsePort(parts[0], 'port');
    return { localPort: port, containerPort: port };
  }
  if (parts.length === 2) {
    return {
      localPort: parsePort(parts[0], 'local port'),
      containerPort: parsePort(parts[1], 'container port'),
    };
  }
  throw new Error(`Invalid port mapping: ${item} (expected 'port' or 'local:container')`);
}

export function parsePortsList(value: string): PortPair[] {
  const items = value
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean);
  if (items.length === 0) {
    throw new Error('No ports specified.');
  }
  return items.map(parsePortPair);
}

export function parseSshConfigContent(content: string): string[] {
  const hosts: string[] = [];
  const lines = content.split(/\r?\n/);
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    const match = line.match(/^[ \t]*[Hh][Oo][Ss][Tt][ \t]+([^#]+)/);
    if (match) {
      const hostPart = match[1].trim();
      const parts = hostPart.split(/[ \t]+/);
      for (const p of parts) {
        const h = p.trim();
        if (h && !h.includes('*') && !h.includes('?')) {
          if (!hosts.includes(h)) {
            hosts.push(h);
          }
        }
      }
    }
  }
  return hosts;
}

export function getSshAliases(configPath?: string): string[] {
  const p = configPath || path.join(os.homedir(), '.ssh', 'config');
  if (!fs.existsSync(p)) return [];
  try {
    const content = fs.readFileSync(p, 'utf8');
    return parseSshConfigContent(content);
  } catch (e) {
    return [];
  }
}

export function portForwardHelp(): string {
  return `devcontainer-cli port-forward — forward host ports to container ports using SSH

Usage:
  devcontainer-cli port-forward [port_mapping] [flags]

Run with no port_mapping for an interactive session: pick any running container
(devcontainers are marked '(devcontainer)'), enter one or more ports, and repeat
for other containers. All tunnels are opened in parallel after you confirm.
Non-devcontainer targets are reached through a devcontainer SSH jump host.

Examples:
  devcontainer-cli port-forward                 # interactive multi-container picker
  devcontainer-cli port-forward 3000
  devcontainer-cli port-forward 8080:80
  devcontainer-cli port-forward 5432:postgres:5432
  devcontainer-cli port-forward 5432 --service postgres
  devcontainer-cli port-forward 3000 --alias my-custom-host

Flags:
  --alias <name>       SSH host alias to use (bypasses auto-discovery)
  --service <name>     Compose service to map port to (default: localhost)
  --no-interactive     Disable interactive prompts (fail on missing config)
  --non-interactive    Disable interactive prompts (fail on missing config)
  -h, --help           Show this help
`;
}

/** SSH alias a container exposes itself under, or null when none is configured. */
function ownSshAlias(containerName: string, aliases: string[]): string | null {
  const candidate = containerWorkspace(containerName) ?? containerName;
  return aliases.includes(candidate) ? candidate : null;
}

/** SSH aliases of running devcontainers, usable as jump hosts. */
function jumpHostAliases(containers: DockerContainer[], aliases: string[]): string[] {
  const result: string[] = [];
  for (const c of containers) {
    if (!c.managed) continue;
    const a = ownSshAlias(c.name, aliases);
    if (a && !result.includes(a)) result.push(a);
  }
  return result;
}

async function pickContainerFromList(
  containers: DockerContainer[],
  prompt: string,
): Promise<DockerContainer> {
  const maxName = Math.max(...containers.map((c) => c.name.length));
  const maxImage = Math.max(...containers.map((c) => c.image.length));
  const chosen = await select(
    'container',
    prompt,
    containers.map((c) => ({
      name: c.name,
      message: `${c.name.padEnd(maxName)}  ${
        c.managed ? chalk.green('(devcontainer)') : chalk.gray('              ')
      }  ${chalk.gray(c.image.padEnd(maxImage))}  ${statusLabel(c)}`,
    })),
  );
  return containers.find((c) => c.name === chosen)!;
}

/**
 * Resolve which SSH alias to tunnel through for a target container, plus the
 * host name the tunnel should reach on the far side. Devcontainers with their
 * own alias are reached directly (targetHost = localhost); everything else is
 * reached by container name through a devcontainer jump host.
 */
async function resolveTarget(
  container: DockerContainer,
  aliases: string[],
  containers: DockerContainer[],
  state: { jumpHost?: string },
  interactive: boolean,
): Promise<{ alias: string; targetHost: string }> {
  if (container.managed) {
    const own = ownSshAlias(container.name, aliases);
    if (own) return { alias: own, targetHost: 'localhost' };
  }

  if (!state.jumpHost) {
    const candidates = jumpHostAliases(containers, aliases);
    if (candidates.length === 1) {
      state.jumpHost = candidates[0];
      console.log(chalk.cyan(`Using SSH jump host '${state.jumpHost}'.`));
    } else if (candidates.length > 1) {
      if (!interactive) {
        throw new Error('Multiple SSH jump hosts available; specify --alias.');
      }
      state.jumpHost = await select(
        'jumpHost',
        'Select the devcontainer SSH host to tunnel through:',
        candidates.map((a) => ({ name: a, message: a })),
      );
    } else {
      if (!interactive) {
        throw new Error(
          `No devcontainer SSH alias found to tunnel through '${container.name}'. Run 'setup-ssh' first or specify --alias.`,
        );
      }
      state.jumpHost = await input(
        'jumpHost',
        `No devcontainer SSH alias found to reach '${container.name}'. Enter SSH alias to tunnel through:`,
        undefined,
        (v) => (v.trim() ? true : 'SSH alias cannot be empty.'),
      );
    }
  }

  return { alias: state.jumpHost, targetHost: container.name };
}

async function buildTunnelsInteractive(flags: PortForwardConfig): Promise<PlannedTunnel[]> {
  const containers = listAllContainers();
  if (containers.length === 0) {
    throw new Error("No running containers found. Start a container with 'devcontainer-cli' first.");
  }

  const aliases = getSshAliases();
  const tunnels: PlannedTunnel[] = [];
  const state: { jumpHost?: string } = { jumpHost: flags.alias };

  for (;;) {
    const container =
      containers.length === 1
        ? containers[0]
        : await pickContainerFromList(containers, 'Select a container to forward from:');
    if (containers.length === 1) {
      console.log(
        chalk.cyan(
          `Using container: ${container.name}${container.managed ? ' (devcontainer)' : ''}`,
        ),
      );
    }

    const { alias, targetHost } = await resolveTarget(
      container,
      aliases,
      containers,
      state,
      flags.interactive,
    );

    const portsStr = await input(
      'ports',
      `Ports to forward from '${container.name}' (e.g. 3000, 8080:80):`,
      undefined,
      (v) => {
        try {
          parsePortsList(v);
          return true;
        } catch (e) {
          return (e as Error).message;
        }
      },
    );

    for (const pair of parsePortsList(portsStr)) {
      tunnels.push({
        localPort: pair.localPort,
        containerPort: pair.containerPort,
        targetHost,
        alias,
        containerName: container.name,
        isDevcontainer: container.managed,
      });
    }

    const more = await confirm('more', 'Add ports from another container?', false);
    if (!more) break;
  }

  return tunnels;
}

function printPlan(tunnels: PlannedTunnel[]): void {
  console.log(chalk.cyan('\nPort forwarding plan:'));
  const maxName = Math.max(...tunnels.map((t) => t.containerName.length));
  for (const t of tunnels) {
    const tag = t.isDevcontainer ? chalk.green('(devcontainer)') : chalk.gray('              ');
    console.log(
      `  ${t.containerName.padEnd(maxName)}  ${tag}  ` +
        chalk.yellow(`localhost:${t.localPort}`) +
        ` → ${t.targetHost}:${t.containerPort}  ` +
        chalk.gray(`(via ${t.alias})`),
    );
  }
  console.log('');
}

function runTunnels(tunnels: PlannedTunnel[]): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const children: ChildProcess[] = [];
    let remaining = tunnels.length;
    let firstError: Error | null = null;
    let terminating = false;
    let settled = false;

    const killAll = () => {
      for (const c of children) {
        if (c.exitCode === null && !c.killed) {
          try {
            c.kill('SIGTERM');
          } catch {
            /* already gone */
          }
        }
      }
    };

    const onSigint = () => {
      terminating = true;
      killAll();
    };
    process.on('SIGINT', onSigint);

    const finish = () => {
      if (settled) return;
      settled = true;
      process.removeListener('SIGINT', onSigint);
      if (firstError) {
        reject(firstError);
      } else {
        console.log(chalk.green(`\nAll SSH tunnels closed.\n`));
        resolve();
      }
    };

    for (const t of tunnels) {
      const child = spawn('ssh', ['-N', '-L', `${t.localPort}:${t.targetHost}:${t.containerPort}`, t.alias], {
        stdio: 'inherit',
      });
      children.push(child);

      child.on('error', (err) => {
        if (!firstError) firstError = new Error(`Failed to spawn SSH process: ${err.message}`);
        terminating = true;
        killAll();
      });

      child.on('close', (code) => {
        remaining--;
        if (code !== 0 && code !== null && !terminating && !firstError) {
          firstError = new Error(
            `SSH tunnel ${t.localPort}→${t.targetHost}:${t.containerPort} (${t.alias}) closed with exit code ${code}`,
          );
          // One failing tunnel tears down the rest so the user isn't left with a
          // partial set silently running in the background.
          terminating = true;
          killAll();
        }
        if (remaining === 0) finish();
      });
    }
  });
}

export async function runPortForward(argv: string[]): Promise<void> {
  const flags = parsePortForwardFlags(argv);
  if (flags.help) {
    console.log(portForwardHelp());
    return;
  }

  // No explicit mapping + interactive → multi-container picker.
  if (!flags.portMapping && flags.interactive) {
    const tunnels = await buildTunnelsInteractive(flags);
    printPlan(tunnels);
    const ok = await confirm(
      'start',
      `Open ${tunnels.length} tunnel${tunnels.length === 1 ? '' : 's'} now?`,
      true,
    );
    if (!ok) {
      console.log(chalk.yellow('Aborted. No tunnels were opened.'));
      return;
    }
    console.log(chalk.green('Press Ctrl+C to terminate the port forwarding session.\n'));
    return runTunnels(tunnels);
  }

  // Explicit single mapping (positional or non-interactive).
  if (!flags.portMapping) {
    throw new Error('Port mapping is required in non-interactive mode.');
  }

  const mapping = parsePortMapping(flags.portMapping, flags.service);

  let alias = flags.alias;
  let containerName = alias ?? mapping.targetHost;
  let isDevcontainer = true;
  if (!alias) {
    const container = await pickManagedContainer('Select devcontainer to forward into:', {
      interactive: flags.interactive,
    });
    containerName = container.name;
    const ws = containerWorkspace(container.name);
    const candidate = ws ?? container.name;
    const aliases = getSshAliases();
    if (aliases.includes(candidate)) {
      alias = candidate;
    } else if (aliases.length === 0) {
      if (!flags.interactive) {
        throw new Error(
          `No SSH aliases found in config. Run 'setup-ssh' first or specify --alias.`,
        );
      }
      alias = await input(
        'alias',
        `No SSH alias found for '${container.name}'. Enter alias manually:`,
        candidate,
        (v) => (v.trim() ? true : 'SSH alias cannot be empty.'),
      );
    } else if (aliases.length === 1) {
      alias = aliases[0];
      console.log(chalk.cyan(`Using SSH alias '${alias}'.`));
    } else {
      if (!flags.interactive) {
        throw new Error(
          `SSH alias '${candidate}' not found in ~/.ssh/config. Specify --alias or run interactively.`,
        );
      }
      alias = await select(
        'alias',
        `Select SSH alias for container '${container.name}':`,
        aliases.map((a) => ({ name: a, message: a })),
      );
    }
  }

  const tunnel: PlannedTunnel = {
    localPort: mapping.localPort,
    containerPort: mapping.containerPort,
    targetHost: mapping.targetHost,
    alias,
    containerName,
    isDevcontainer,
  };

  printPlan([tunnel]);
  console.log(chalk.green('Press Ctrl+C to terminate the port forwarding session.\n'));
  return runTunnels([tunnel]);
}

File: cli/internal/commands/port_forward.go
package commands

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"github.com/fatih/color"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/infra/containerpicker"
	"github.com/joacohbc/my-devcontainer-installer/cli/internal/infra/prompt"
	"github.com/spf13/cobra"
)

func init() { register(newPortForwardCommand()) }

func newPortForwardCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "port-forward [port_mapping]",
		Short: "Forward host port to a container port using SSH",
		Long: `devcontainer-cli port-forward — forward host ports to container ports using SSH

Run with no port_mapping for an interactive session: pick any running container
(devcontainers are marked '(devcontainer)'), enter one or more ports, and repeat
for other containers. All tunnels are opened in parallel after you confirm.

Examples:
  devcontainer-cli port-forward                 # interactive multi-container picker
  devcontainer-cli port-forward 3000
  devcontainer-cli port-forward 8080:80
  devcontainer-cli port-forward 5432:postgres:5432
  devcontainer-cli port-forward 5432 --service postgres
  devcontainer-cli port-forward 3000 --alias my-custom-host`,
		Args:         cobra.MaximumNArgs(1),
		SilenceUsage: true,
		RunE:         runPortForward,
	}
	cmd.Flags().String("alias", "", "SSH host alias to use (bypasses auto-discovery)")
	cmd.Flags().String("service", "", "Compose service to map port to (default: localhost)")
	cmd.Flags().Bool("no-interactive", false, "Disable interactive prompts (fail on missing config)")
	cmd.Flags().Bool("non-interactive", false, "Disable interactive prompts (fail on missing config)")
	return cmd
}

type parsedMapping struct {
	localPort     int
	targetHost    string
	containerPort int
}

type portPair struct {
	localPort     int
	containerPort int
}

type plannedTunnel struct {
	localPort      int
	containerPort  int
	targetHost     string
	alias          string
	containerName  string
	isDevcontainer bool
}

func parsePort(value, label string) (int, error) {
	n, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil || n <= 0 || n > 65535 {
		return 0, fmt.Errorf("invalid %s: %s", label, value)
	}
	return n, nil
}

func parsePortMapping(mapping, defaultService string) (parsedMapping, error) {
	parts := strings.Split(mapping, ":")
	for i := range parts {
		parts[i] = strings.TrimSpace(parts[i])
	}
	switch len(parts) {
	case 1:
		port, err := parsePort(parts[0], "port")
		if err != nil {
			return parsedMapping{}, err
		}
		host := defaultService
		if host == "" {
			host = "localhost"
		}
		return parsedMapping{localPort: port, targetHost: host, containerPort: port}, nil
	case 2:
		local, err := parsePort(parts[0], "local port")
		if err != nil {
			return parsedMapping{}, err
		}
		container, err := parsePort(parts[1], "container port")
		if err != nil {
			return parsedMapping{}, err
		}
		host := defaultService
		if host == "" {
			host = "localhost"
		}
		return parsedMapping{localPort: local, targetHost: host, containerPort: container}, nil
	case 3:
		local, err := parsePort(parts[0], "local port")
		if err != nil {
			return parsedMapping{}, err
		}
		targetHost := parts[1]
		if targetHost == "" {
			return parsedMapping{}, fmt.Errorf("invalid target host: empty string")
		}
		container, err := parsePort(parts[2], "container port")
		if err != nil {
			return parsedMapping{}, err
		}
		if defaultService != "" && defaultService != targetHost {
			return parsedMapping{}, fmt.Errorf("conflicting target hosts: mapping specifies '%s' but --service is '%s'", targetHost, defaultService)
		}
		return parsedMapping{localPort: local, targetHost: targetHost, containerPort: container}, nil
	default:
		return parsedMapping{}, fmt.Errorf("invalid port mapping format: %s", mapping)
	}
}

func parsePortPair(item string) (portPair, error) {
	parts := strings.Split(item, ":")
	for i := range parts {
		parts[i] = strings.TrimSpace(parts[i])
	}
	switch len(parts) {
	case 1:
		p, err := parsePort(parts[0], "port")
		if err != nil {
			return portPair{}, err
		}
		return portPair{localPort: p, containerPort: p}, nil
	case 2:
		local, err := parsePort(parts[0], "local port")
		if err != nil {
			return portPair{}, err
		}
		container, err := parsePort(parts[1], "container port")
		if err != nil {
			return portPair{}, err
		}
		return portPair{localPort: local, containerPort: container}, nil
	default:
		return portPair{}, fmt.Errorf("invalid port mapping: %s (expected 'port' or 'local:container')", item)
	}
}

func parsePortsList(value string) ([]portPair, error) {
	var pairs []portPair
	for _, item := range strings.Split(value, ",") {
		if t := strings.TrimSpace(item); t != "" {
			p, err := parsePortPair(t)
			if err != nil {
				return nil, err
			}
			pairs = append(pairs, p)
		}
	}
	if len(pairs) == 0 {
		return nil, fmt.Errorf("no ports specified")
	}
	return pairs, nil
}

var hostLineRe = regexp.MustCompile(`(?i)^[ \t]*host[ \t]+([^#]+)`)

func parseSSHConfigContent(content string) []string {
	var hosts []string
	seen := map[string]bool{}
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		m := hostLineRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		for _, p := range strings.Fields(strings.TrimSpace(m[1])) {
			if p != "" && !strings.Contains(p, "*") && !strings.Contains(p, "?") && !seen[p] {
				seen[p] = true
				hosts = append(hosts, p)
			}
		}
	}
	return hosts
}

func getSSHAliases() []string {
	home, _ := os.UserHomeDir()
	p := filepath.Join(home, ".ssh", "config")
	data, err := os.ReadFile(p)
	if err != nil {
		return nil
	}
	return parseSSHConfigContent(string(data))
}

func ownSSHAlias(containerName string, aliases []string) string {
	candidate := containerpicker.ContainerWorkspace(containerName)
	if candidate == "" {
		candidate = containerName
	}
	for _, a := range aliases {
		if a == candidate {
			return candidate
		}
	}
	return ""
}

func jumpHostAliases(containers []containerpicker.Container, aliases []string) []string {
	var result []string
	seen := map[string]bool{}
	for _, c := range containers {
		if !c.Managed {
			continue
		}
		if a := ownSSHAlias(c.Name, aliases); a != "" && !seen[a] {
			seen[a] = true
			result = append(result, a)
		}
	}
	return result
}

func pickContainerFromList(containers []containerpicker.Container, message string) (containerpicker.Container, error) {
	choices := make([]prompt.Choice, len(containers))
	for i, c := range containers {
		tag := "              "
		if c.Managed {
			tag = color.GreenString("(devcontainer)")
		}
		choices[i] = prompt.Choice{Value: c.Name, Label: fmt.Sprintf("%s  %s  %s  %s", c.Name, tag, color.WhiteString(c.Image), containerpicker.StatusLabel(c))}
	}
	chosen, err := prompt.Select(message, choices, choices[0].Value)
	if err != nil {
		return containerpicker.Container{}, err
	}
	for _, c := range containers {
		if c.Name == chosen {
			return c, nil
		}
	}
	return containers[0], nil
}

func resolveTarget(container containerpicker.Container, aliases []string, containers []containerpicker.Container, jumpHost *string, interactive bool) (alias, targetHost string, err error) {
	if container.Managed {
		if own := ownSSHAlias(container.Name, aliases); own != "" {
			return own, "localhost", nil
		}
	}
	if *jumpHost == "" {
		candidates := jumpHostAliases(containers, aliases)
		switch {
		case len(candidates) == 1:
			*jumpHost = candidates[0]
			color.Cyan("Using SSH jump host '%s'.", *jumpHost)
		case len(candidates) > 1:
			if !interactive {
				return "", "", fmt.Errorf("multiple SSH jump hosts available; specify --alias")
			}
			choices := make([]prompt.Choice, len(candidates))
			for i, a := range candidates {
				choices[i] = prompt.Choice{Value: a, Label: a}
			}
			sel, serr := prompt.Select("Select the devcontainer SSH host to tunnel through:", choices, candidates[0])
			if serr != nil {
				return "", "", serr
			}
			*jumpHost = sel
		default:
			if !interactive {
				return "", "", fmt.Errorf("no devcontainer SSH alias found to tunnel through '%s'. Run 'setup-ssh' first or specify --alias", container.Name)
			}
			in, ierr := prompt.Input(fmt.Sprintf("No devcontainer SSH alias found to reach '%s'. Enter SSH alias to tunnel through:", container.Name), "", func(v string) error {
				if strings.TrimSpace(v) == "" {
					return fmt.Errorf("SSH alias cannot be empty")
				}
				return nil
			})
			if ierr != nil {
				return "", "", ierr
			}
			*jumpHost = in
		}
	}
	return *jumpHost, container.Name, nil
}

func buildTunnelsInteractive(flagAlias string, interactive bool) ([]plannedTunnel, error) {
	containers := containerpicker.ListAll()
	if len(containers) == 0 {
		return nil, fmt.Errorf("no running containers found. Start a container with 'devcontainer-cli' first")
	}
	aliases := getSSHAliases()
	var tunnels []plannedTunnel
	jumpHost := flagAlias

	for {
		var container containerpicker.Container
		var err error
		if len(containers) == 1 {
			container = containers[0]
			tag := ""
			if container.Managed {
				tag = " (devcontainer)"
			}
			color.Cyan("Using container: %s%s", container.Name, tag)
		} else {
			container, err = pickContainerFromList(containers, "Select a container to forward from:")
			if err != nil {
				return nil, err
			}
		}

		alias, targetHost, terr := resolveTarget(container, aliases, containers, &jumpHost, interactive)
		if terr != nil {
			return nil, terr
		}

		portsStr, ierr := prompt.Input(fmt.Sprintf("Ports to forward from '%s' (e.g. 3000, 8080:80):", container.Name), "", func(v string) error {
			_, e := parsePortsList(v)
			return e
		})
		if ierr != nil {
			return nil, ierr
		}
		pairs, _ := parsePortsList(portsStr)
		for _, pair := range pairs {
			tunnels = append(tunnels, plannedTunnel{
				localPort:      pair.localPort,
				containerPort:  pair.containerPort,
				targetHost:     targetHost,
				alias:          alias,
				containerName:  container.Name,
				isDevcontainer: container.Managed,
			})
		}

		more, merr := prompt.Confirm("Add ports from another container?", false)
		if merr != nil {
			return nil, merr
		}
		if !more {
			break
		}
	}
	return tunnels, nil
}

func printTunnelPlan(tunnels []plannedTunnel) {
	color.Cyan("\nPort forwarding plan:")
	for _, t := range tunnels {
		tag := "              "
		if t.isDevcontainer {
			tag = color.GreenString("(devcontainer)")
		}
		fmt.Printf("  %s  %s  %s → %s:%d  %s\n",
			t.containerName, tag,
			color.YellowString("localhost:%d", t.localPort),
			t.targetHost, t.containerPort,
			color.WhiteString("(via %s)", t.alias))
	}
	fmt.Println()
}

func runTunnels(tunnels []plannedTunnel) error {
	var wg sync.WaitGroup
	var mu sync.Mutex
	var firstErr error
	cmds := make([]*exec.Cmd, 0, len(tunnels))

	killAll := func() {
		for _, c := range cmds {
			if c.Process != nil {
				_ = c.Process.Kill()
			}
		}
	}

	for _, t := range tunnels {
		c := exec.Command("ssh", "-N", "-L", fmt.Sprintf("%d:%s:%d", t.localPort, t.targetHost, t.containerPort), t.alias)
		c.Stdin = os.Stdin
		c.Stdout = os.Stdout
		c.Stderr = os.Stderr
		if err := c.Start(); err != nil {
			mu.Lock()
			if firstErr == nil {
				firstErr = fmt.Errorf("failed to spawn SSH process: %w", err)
			}
			mu.Unlock()
			killAll()
			continue
		}
		cmds = append(cmds, c)
		tun := t
		wg.Add(1)
		go func(cmd *exec.Cmd) {
			defer wg.Done()
			err := cmd.Wait()
			if err != nil {
				mu.Lock()
				if firstErr == nil {
					firstErr = fmt.Errorf("SSH tunnel %d→%s:%d (%s) closed: %v", tun.localPort, tun.targetHost, tun.containerPort, tun.alias, err)
				}
				mu.Unlock()
				killAll()
			}
		}(c)
	}

	wg.Wait()
	if firstErr != nil {
		return firstErr
	}
	color.Green("\nAll SSH tunnels closed.\n")
	return nil
}

func runPortForward(cmd *cobra.Command, args []string) error {
	alias, _ := cmd.Flags().GetString("alias")
	service, _ := cmd.Flags().GetString("service")
	noI, _ := cmd.Flags().GetBool("no-interactive")
	nonI, _ := cmd.Flags().GetBool("non-interactive")
	interactive := !(noI || nonI)

	var portMapping string
	if len(args) == 1 {
		portMapping = args[0]
	}

	if portMapping == "" && interactive {
		tunnels, err := buildTunnelsInteractive(alias, interactive)
		if err != nil {
			return err
		}
		printTunnelPlan(tunnels)
		ok, cerr := prompt.Confirm(fmt.Sprintf("Open %d tunnel(s) now?", len(tunnels)), true)
		if cerr != nil {
			return cerr
		}
		if !ok {
			color.Yellow("Aborted. No tunnels were opened.")
			return nil
		}
		color.Green("Press Ctrl+C to terminate the port forwarding session.\n")
		return runTunnels(tunnels)
	}

	if portMapping == "" {
		return fmt.Errorf("port mapping is required in non-interactive mode")
	}

	mapping, err := parsePortMapping(portMapping, service)
	if err != nil {
		return err
	}

	containerName := alias
	if containerName == "" {
		containerName = mapping.targetHost
	}
	if alias == "" {
		container, perr := containerpicker.PickManaged("Select devcontainer to forward into:", containerpicker.PickOptions{Interactive: interactive})
		if perr != nil {
			return perr
		}
		containerName = container.Name
		candidate := containerpicker.ContainerWorkspace(container.Name)
		if candidate == "" {
			candidate = container.Name
		}
		aliases := getSSHAliases()
		switch {
		case contains(aliases, candidate):
			alias = candidate
		case len(aliases) == 0:
			if !interactive {
				return fmt.Errorf("no SSH aliases found in config. Run 'setup-ssh' first or specify --alias")
			}
			alias, err = prompt.Input(fmt.Sprintf("No SSH alias found for '%s'. Enter alias manually:", container.Name), candidate, func(v string) error {
				if strings.TrimSpace(v) == "" {
					return fmt.Errorf("SSH alias cannot be empty")
				}
				return nil
			})
			if err != nil {
				return err
			}
		case len(aliases) == 1:
			alias = aliases[0]
			color.Cyan("Using SSH alias '%s'.", alias)
		default:
			if !interactive {
				return fmt.Errorf("SSH alias '%s' not found in ~/.ssh/config. Specify --alias or run interactively", candidate)
			}
			choices := make([]prompt.Choice, len(aliases))
			for i, a := range aliases {
				choices[i] = prompt.Choice{Value: a, Label: a}
			}
			alias, err = prompt.Select(fmt.Sprintf("Select SSH alias for container '%s':", container.Name), choices, aliases[0])
			if err != nil {
				return err
			}
		}
	}

	tunnel := plannedTunnel{
		localPort:      mapping.localPort,
		containerPort:  mapping.containerPort,
		targetHost:     mapping.targetHost,
		alias:          alias,
		containerName:  containerName,
		isDevcontainer: true,
	}
	printTunnelPlan([]plannedTunnel{tunnel})
	color.Green("Press Ctrl+C to terminate the port forwarding session.\n")
	return runTunnels([]plannedTunnel{tunnel})
}

func contains(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}
